nni/retiarii/hub/pytorch/autoformer.py

```diff
@@ -146,91 +135,112 @@ class TransformerEncoderLayer(nn.Module):
-            embed_dim=embed_dim,
-            fixed_embed_dim=fixed_embed_dim,
-            num_heads=num_heads,
-            attn_drop=attn_drop,
+            embed_dim = embed_dim,
+            num_heads = num_heads,
+            attn_drop = attn_drop,
             proj_drop=proj_drop,
             rpe=rpe,
             qkv_bias=qkv_bias,
             qk_scale=qk_scale,
-            rpe_length=rpe_length)
+            rpe_length=rpe_length,
+            head_dim = head_dim
+        )
 
         self.attn_layer_norm = nn.LayerNorm(embed_dim)
         self.ffn_layer_norm = nn.LayerNorm(embed_dim)
+
         self.activation_fn = nn.GELU()
+
         self.fc1 = nn.Linear(
-            cast(int, embed_dim), cast(int, nn.ValueChoice.to_int(
-                embed_dim * mlp_ratio)))
+            cast(int, embed_dim), 
+            cast(int, nn.ValueChoice.to_int(embed_dim * mlp_ratio))
+        )
         self.fc2 = nn.Linear(
-            cast(int, nn.ValueChoice.to_int(
-                embed_dim * mlp_ratio)),
-            cast(int, embed_dim))
+            cast(int, nn.ValueChoice.to_int(embed_dim * mlp_ratio)),
+            cast(int, embed_dim)
+        )
+
+    def maybe_layer_norm(self, layer_norm, x, before=False, after=False):
+        assert before ^ after
+        if after ^ self.normalize_before:
+            return layer_norm(x)
+        else:
+            return x
 
     def forward(self, x):
         """
         Args:
             x (Tensor): input to the layer of shape `(batch, patch_num , sample_embed_dim)`
-
         Returns:
             encoded output of shape `(batch, patch_num, sample_embed_dim)`
         """
         residual = x
         x = self.maybe_layer_norm(self.attn_layer_norm, x, before=True)
         x = self.attn(x)
-        x = nn.functional.dropout(x, p=self.dropout, training=self.training)
+        x = F.dropout(x, p=self.dropout, training=self.training)
         x = self.drop_path(x)
         x = residual + x
         x = self.maybe_layer_norm(self.attn_layer_norm, x, after=True)
-
+        
         residual = x
         x = self.maybe_layer_norm(self.ffn_layer_norm, x, before=True)
         x = self.fc1(x)
         x = self.activation_fn(x)
-        x = nn.functional.dropout(x, p=self.dropout, training=self.training)
+        x = F.dropout(x, p=self.dropout, training=self.training)
         x = self.fc2(x)
-        x = nn.functional.dropout(x, p=self.dropout, training=self.training)
+        x = F.dropout(x, p=self.dropout, training=self.training)
         x = self.drop_path(x)
         x = residual + x
         x = self.maybe_layer_norm(self.ffn_layer_norm, x, after=True)
+
         return x
 
-    def maybe_layer_norm(self, layer_norm, x, before=False, after=False):
-        assert before ^ after
-        if after ^ self.normalize_before:
-            return layer_norm(x)
-        else:
-            return x
 
+class ClsToken(nn.Module):
+    """ Concat class token before patch embedding. 
```

Comment:
Suggest adding more explanations here, about how forward works (slicing the token and etc.).

Actually a more elegant way is to make this operation a basic unit, and create a `MixedClsToken` as one of the supernet operation, like this:

```python
@basic_unit
class ClsToken(...):
  def forward(self, x):
    # this forward is only used in fixed architecture
    return torch.cat((self.cls_token.expand(x.shape[0], -1, -1), x), 1)

class MixedClsToken(...):
   def forward(
```

Note that both of these operations need to be put in this `autoformer.py` because ClsToken is not considered as a general operation.

When running the strategy, you need to register `MixedClsToken` as a extra mutation hook, like this:

https://github.com/microsoft/nni/blob/570448eab83939008ac6a2d8b4ba6ed85653328a/test/algo/nas/test_space_hub_oneshot.py#L78

Reply:
```
@basic_unit
class ClsToken(nn.Module):
    """ Concat class token with dim=embed_dim before patch embedding. 
    """
    def __init__(self, embed_dim: int):
        super().__init__()
        self.cls_token = nn.Parameter(torch.zeros(1, 1, embed_dim))
        trunc_normal_(self.cls_token, std=.02)
    
    def forward(self, x):
        return torch.cat((self.cls_token.expand(x.shape[0], -1, -1), x), dim=1)


class MixedClsToken(MixedOperation, ClsToken):
    """ Mixed class token concat operation.

    Supported arguments are:

    - ``embed_dim``
    
    Prefix of cls_token will be sliced.
    """
    bound_type = ClsToken
    argument_list = ['embed_dim']

    def super_init_argument(self, name: str, value_choice: ValueChoiceX):
        return max(traverse_all_options(value_choice))

    def forward_with_args(self, embed_dim, 
                        inputs: torch.Tensor) -> torch.Tensor:
        cls_token = self.cls_token[..., :embed_dim]

        return torch.cat((cls_token.expand(inputs.shape[0], -1, -1), inputs), dim=1)
```
when i use `super().__init__()`, everything is all right, but when I use `super(ClsToken, self).__init__()` rather than `super().__init__()` in `ClsToken`, there raise exception as follow.

```
  File "nni/nni/common/serializer.py", line 508, in __init__
    super().__init__(symbol=base, args=args, kwargs=kwargs, call_super=call_super)
  File "nni/nni/common/serializer.py", line 133, in __init__
    super().__init__(
  File "nni/nni/retiarii/hub/pytorch/autoformer.py", line 219, in __init__
    super(ClsToken, self).__init__()
TypeError: __init__() missing 3 required positional arguments: 'symbol', 'args', and 'kwargs'
```